lib/steep/ast/types/intersection.rb

```diff
@@ -43,7 +43,8 @@ def self.build(types:, location: nil)
         end
 
         def ==(other)
-          other.is_a?(Intersection) && other.types == types
+          (other.is_a?(Intersection) && other.types == types) ||
```

Comment:
The `==` on types are to test the equality of the values. You should use `Subtyping::Check` to test semantic relation between types.